### PR TITLE
Remove assumption from `tiles()` method that tile rows/cols are ordered…

### PR DESCRIPTION
**Pull request by fsvenson**
… west-east and north-south

A TileMatrixSet created for a projection like an S2 face does not comply with this assumption.

## What I am changing

The `tiles()` method did not return the expected tiles overlapping a bbox for a tile matrix created using an S2 face projection (see [this](https://s2geometry.io/resources/earthcube)).

My change removes the assumption that the bounds of the tile matrix set must be aligned such that the top-left corner corresponds to the geographic north-west etc.

## How I did it
Simply check the tile indices of the north-west and south-east tiles and iterate according to which is greater.

## How you can test it
One can test using a TileMatrixSet created like this:
```python
crs = CRS.from_proj4(f"+proj=s2 +lat_0=0.0 +lon_0=-90.0 +ellps=WGS84 +UVtoST=quadratic")
extent = [0.0, 0.0, 1.0, 1.0]
s2f4 = TileMatrixSet.custom(extent, crs, identifier="S2F4")
overlapping_tiles = s2f4.tiles(-100, 27, -95, 33, [6])
assert len(overlapping_tiles) > 1 
```

### CODE REVIEW

File: morecantile/models.py [hunk 1]
```diff
@@ -778,13 +778,15 @@ def tiles(
...
                 )  # Not in mercantile
-                lr_tile = self.tile(e - LL_EPSILON, s + LL_EPSILON, z)
+                se_tile = self.tile(e - LL_EPSILON, s + LL_EPSILON, z)
+                minx, maxx = (nw_tile.x, se_tile.x) if nw_tile.x < se_tile.x else (se_tile.x, nw_tile.x)
+                miny, maxy = (nw_tile.y, se_tile.y) if nw_tile.y < se_tile.y else (se_tile.y, nw_tile.y)
 
-                for i in range(ul_tile.x, lr_tile.x + 1):
-                    for j in range(ul_tile.y, lr_tile.y + 1):
+                for i in range(minx, maxx + 1):
+                    for j in range(miny, maxy + 1):
```
Comment (vincentsarago) [inline morecantile/models.py:789]: FYI: This originally comes from `mercantile` https://github.com/mapbox/mercantile/blob/main/mercantile/__init__.py#L541-L547 

Could you add a tests for this? (like something that was failing with previous version 🙏 )

Comment (fsvenson) [reply morecantile/models.py:797]: Got it! I guess it makes sense for `mercantile` to only handle the standard axis ordering.

Test added.

**Comment by vincentsarago** [comment]
I can take it from here 🙏 

thanks for starting the PR, it's really interesting that you are able to use S2 with morecantile 👀

**Comment by fsvenson** [comment]
Thanks! Yes, my idea was that it should work well since each side of the cube is a quad tree.

I might come back with another pull request for the `feature()` method so that it returns a densified polygon, representing the projected tile, instead of just the bounding box in lat/lon, which obviously does not work well for S2.

**Pull request merged**